src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildUpstreamNodeAction.java

```diff
@@ -0,0 +1,53 @@
+package org.jenkinsci.plugins.workflow.support.steps.build;
+
+import hudson.model.InvisibleAction;
+import hudson.model.Run;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import java.util.Objects;
+
+/**
+ * Attached to newly-created builds in order to point back to the triggering FlowNode.
+ *
+ * We annotate the downstream build instead of upstream in order to support no-wait builds,
+ * because the downstream run id is not available when they're still queued.
+ *
+ * Needed for Blue Ocean to annotate the correct step.
+ * @see: https://issues.jenkins-ci.org/browse/JENKINS-38339
+ */
+public class BuildUpstreamNodeAction extends InvisibleAction {
+
+    private final String upstreamNodeId;
+    private final String upstreamRunId;
+
+    public BuildUpstreamNodeAction(FlowNode node, Run<?, ?> invokingRun) {
+        this.upstreamNodeId = node.getId();
+        this.upstreamRunId = invokingRun.getExternalizableId();
+    }
+
+    public String getUpstreamNodeId() {
+        return upstreamNodeId;
+    }
+
+    public String getUpstreamRunId() {
+        return upstreamRunId;
+    }
+
```

Comment:
Might want to add a convenience method to return the actual upstream `FlowNode`, if the upstream build still exists etc. Depends on what your downstream PR looks like.

Reply:
Wouldn't really help much in this case, as BO doesn't actually return all FlowNodes back via REST, and we have to walk the graph a little to annotate the node that makes it out to the UI. And that code may change once I've talked to Vivek about it, because I have a suspicion it (the BO code) is not the best way to do what we need.